Document GUI practices (#2215)
knausj85/knausj_talon: PRACTICES.md

- Use the `sleep` action instead of the Python `time.sleep` function.
- Avoid sleeping inside a `cron` callback because public Talon runs all such callbacks in a single thread.
- Avoid command or action implementations that sleep for more than a total of ~200 ms. Talon is unresponsive to commands while sleeping; long delays also trigger watchdog warnings in the Talon log. Consider using `cron` to schedule actions requiring a longer delay.

## GUI Practices

- Please add a screenshot showing GUI changes to the PR description to make review easier.

> - Use a `tag` if you want to do context matching on if a GUI is showing.

Reviewer: ```suggestion
- Manually activate a `tag` if you want a context to match whether a GUI element is visible.
```

### imgui Practices

> - Keep operations in an imgui drawing function efficient and limited because the imgui framework will frequently and repeatedly call the function.

Reviewer: ```suggestion
- Keep operations in an imgui drawing function (decorated by `imgui.open`) efficient and limited because the imgui framework will frequently and repeatedly call the function. For example, cache displayed information in a global variable.
```

> - Create a button and command for closing the imgui. Make the button text show the command's spoken form.

Reviewer: ```suggestion
- Create a button and voice command for closing each imgui window.
```

> - When listing options, provide buttons and a command for choosing the options. Either make the buttons show the exact spoken forms for selecting the corresponding options or explain the spoken forms in the imgui.

Reviewer: ```suggestion
- For each option in a list of options, provide a button and voice command. Either make the buttons show the exact spoken forms for selecting the corresponding options or explain the spoken forms elsewhere in the UI.
```

> - Keep the new user message's description of Community GUIs consistent with new GUIs. Update the description if you have a good reason to not follow the patterns it documents.

Reviewer: ```suggestion
- Keep the "new user message" description consistent with new UI. Update the description if you have a good reason to not follow the patterns it documents.
```

> - If an imgui sometimes goes off the screen from having too much height, consider using pagination like the help system does.

Reviewer: ```suggestion
- If your imgui window is too tall for the some screens, consider using pagination buttons like the help system.
```

> - If your imgui window is too tall for the some screens, consider using pagination buttons like the help system.

Reviewer: ```suggestion
- If your imgui window is too tall for some screens, consider using pagination buttons like the help system.
```

> - Show spoken forms with Talonscript syntax for things like optional words or alternative spoken forms.

Reviewer: ```suggestion
- Try to avoid having multiple spoken forms for the same button, but if you must do this, show alternatives using TalonScript syntax (e.g. `(foo | bar) [baz]`).
```

> - Try to avoid having multiple spoken forms for the same button, but if you must do this, show alternatives using TalonScript syntax (e.g. `(foo | bar) [baz]`).

Reviewer: ```suggestion
- Try to avoid having multiple spoken forms for the same button. If you must, show alternatives using TalonScript syntax (e.g. `(foo | bar) [baz]`).
```
